// deppack/lib/deppack.js
// Generated by CoffeeScript 1.9.1
var browserResolve, detective, each, fs, getHeader, getModuleRootName, getModuleRootPath, loadFile, newlinesIn, nlre, packDeps, readFile, readFileAndProcess, requireDefinition, shims, sysPath;

sysPath = require('path');

fs = require('fs');

detective = require('detective');

browserResolve = require('browser-resolve');

requireDefinition = require('commonjs-require-definition');

each = require('async-each');

shims = ['assert', 'buffer', 'child_process', 'cluster', 'crypto', 'dgram', 'dns', 'events', 'fs', 'http', 'https', 'net', 'os', 'path', 'punycode', 'querystring', 'readline', 'repl', 'string_decoder', 'tls', 'tty', 'url', 'util', 'vm', 'zlib'];

readFile = function(path, callback) {
  return fs.readFile(path, {
    encoding: 'utf8'
  }, callback);
};

nlre = /\n/g;

newlinesIn = function(src) {
  var newlines;
  if (!src) {
    return 0;
  }
  newlines = src.match(nlre);
  if (newlines) {
    return newlines.length;
  } else {
    return 0;
  }
};

getModuleRootPath = function(path) {
  var index, split;
  split = path.split('/');
  index = split.lastIndexOf('node_modules');
  return split.slice(0, index + 2).join('/');
};

getModuleRootName = function(path) {
  var index, split;
  split = path.split('/');
  index = split.lastIndexOf('node_modules');
  return split[index + 1];
};

getHeader = function(moduleName) {
  return "require.register('" + moduleName + "', function (exp, req, mod) {\nmod.exports = (function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require==\"function\"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error(\"Cannot find module '\"+o+\"'\");throw f.code=\"MODULE_NOT_FOUND\",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require==\"function\"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s}) ({";
};

readFileAndProcess = function(filePath, json, paths, rollback, allFiles, stop, done, loadDeps, pid) {
  return readFile(filePath, function(err, src) {
    var deps, getResult, item, itemHandler, resolved;
    if (err && rollback) {
      return stop(err);
    }
    deps = detective(src);
    resolved = {};
    item = {
      id: filePath,
      filename: filePath,
      paths: paths,
      "package": json
    };
    getResult = function() {
      return {
        id: filePath,
        source: src,
        deps: resolved,
        file: filePath
      };
    };
    if (deps.length === 0) {
      allFiles[filePath] = getResult();
      return done();
    }
    itemHandler = function(dep, cb) {
      return browserResolve(dep, item, function(err, fullPath) {
        if (err) {
          return cb(err);
        }
        resolved[dep] = fullPath;
        return cb(null, fullPath);
      });
    };
    each(deps, itemHandler, function(err, fullPathDeps) {
      if (err) {
        if (rollback) {
          return stop(null, '');
        } else {
          return stop(err);
        }
      }
      allFiles[filePath] = getResult();
      return fullPathDeps.filter(function(filePath) {
        return filePath;
      }).forEach(function(filePath) {
        return loadDeps(filePath, pid);
      });
    });
  });
};

packDeps = function(modulePath, header, deps, ignoreRequireDefinition) {
  var entries, str, stringDeps;
  entries = [];
  stringDeps = deps.map(function(dep) {
    if (dep.entry && dep.order) {
      entries[dep.order] = dep.id;
    } else if (dep.entry) {
      entries.push(row.id);
    }
    deps = Object.keys(dep.deps || {}).sort().map(function(key) {
      return (JSON.stringify(key)) + ":" + (JSON.stringify(dep.deps[key]));
    });
    return [JSON.stringify(dep.id), ':[', 'function(require,module,exports){\n', dep.source, '\n},', "{ " + (deps.join(',')) + " }", ']'].join('');
  });
  entries = entries.filter(function(x) {
    return x;
  });
  str = '';
  if (!ignoreRequireDefinition) {
    str += requireDefinition;
  }
  str += header += stringDeps.join(',');
  str += '},{},' + JSON.stringify(entries) + (")('" + modulePath + "'); }) \n ");
  return str;
};

loadFile = function(filePath, opts, callback) {
  var allFiles, entryModuleFilePath, loadDeps, paths, ref, stop, stopped, streams;
  allFiles = {};
  streams = {};
  stopped = false;
  if (typeof opts === 'function') {
    callback = opts;
    opts = null;
  }
  if (opts == null) {
    opts = {};
  }
  if (opts.paths == null) {
    opts.paths = ((ref = process.env.NODE_PATH) != null ? ref.split(':') : void 0) || [];
  }
  if (opts.basedir == null) {
    opts.basedir = process.cwd();
  }
  if (Array.isArray(opts.shims)) {
    shims = shims.concat(opts.shims);
  }
  paths = opts.paths.map(function(path) {
    return sysPath.resolve(opts.basedir, path);
  });
  filePath = sysPath.resolve(opts.basedir, filePath);
  entryModuleFilePath = filePath;
  stop = function(error, data) {
    stopped = true;
    return callback(error, data);
  };
  loadDeps = function(filePath, parentId) {
    var depRootPath, done, error, json, pid;
    if (stopped) {
      return;
    }
    depRootPath = opts.rootPath || getModuleRootPath(filePath);
    pid = Date.now() + Math.round(Math.random() * 1000000);
    streams[pid] = false;
    if (parentId) {
      delete streams[parentId];
    }
    done = function() {
      var deps, header, packed;
      delete streams[pid];
      if (Object.keys(streams).length !== 0) {
        return;
      }
      deps = Object.keys(allFiles).map(function(key) {
        return allFiles[key];
      });
      header = opts.header || getHeader(opts.name || getModuleRootName(entryModuleFilePath));
      packed = packDeps(filePath, header, deps, opts.ignoreRequireDefinition);
      return callback(null, packed);
    };
    try {
      json = require(sysPath.join(depRootPath, 'package.json'));
    } catch (_error) {
      error = _error;
      return done();
    }
    if (allFiles[filePath]) {
      return done();
    }
    return readFileAndProcess(filePath, json, paths, opts.rollback, allFiles, stop, done, loadDeps, pid);
  };
  return loadDeps(filePath);
};

module.exports = loadFile;
